### PR TITLE
Vault Strategy

### PR DESCRIPTION
**Pull request by Defcon023**
## Single Strategy Vault
Closes : #194  
### **1. Functionality Overview:**
- The Base Strategy module manages deposits, withdrawals, shares, and profit and loss accounting.

### **2. Deposit (Mints) Process:**
- Users can deposit `ASSET` tokens to get strategy shares.

### **3. Withdrawal (Redeems) Process:**
- Users can withdraw/redeem their shares provided there's available liquidity.
- To ensure compatibility with the ERC-4626 standard and manage potential losses:
  - Both `withdraw` and `redeem` functions incorporate an optional `maxLoss` parameter.
  - The default for `maxLoss` in `withdraw` is set to 0 (i.e., the function will revert if there's any loss).
  - The default for `maxLoss` in `redeem` is set to 10_000 (or 100%).

### **4. Strategy Shares:**
- Upon depositing, the strategy allots shares to depositors to denote their proportional asset share.
- These shares are ERC20 compliant, transferable, and yield-bearing tokens (conforming to ERC4626).

### **5. Profit & Loss Accounting:**
- The strategy assesses profit and losses derived from its yield-generating ventures.
  - Evaluation is based on the comparison of the current `totalAssets` of the strategy with the value returned from `_harvestAndReport()`.
  - If `totalAssets` is less than `newTotalAssets`, a profit is recorded.
  - Conversely, if `totalAssets` is more than `newTotalAssets`, a loss is recorded.
  - Both profit and loss scenarios affect the strategy's `totalAssets`.

### **6. Profit Distribution:**
- Profit from report actions is stored in a buffer.
- This buffer undergoes a linear unlocking over a predefined period at `profitUnlockingRate`.
- Profits have a maximum locking time represented by `profitMaxUnlockTime` (in seconds). The distribution is done gradually.
- To manage gas expenses during profit unlocking, the locking time for a profit is determined as a weighted average of the new profit and the previous one. The formulae are as follows:
  - `new_locking_period = (current_locked_profit * pending_time_to_unlock + new_profit * PROFIT_MAX_UNLOCK_TIME) / (current_locked_profit + new_profit)`
  - `new_profit_unlocking_rate = (locked_profit + new_profit) / new_locking_period`
- In case of losses, they are compensated using the locked profit if available.

### CODE REVIEW

File: src/ReserveVault/strategy/AaveV3Strategy.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "../BaseStrategy.sol";
+import "./IPool.sol";
+
+
+contract AaveV3Strategy is BaseStrategy {
```
Comment (samtvlabs) [inline src/ReserveVault/strategy/AaveV3Strategy.sol:9]: Natspec

File: src/ReserveVault/strategy/AaveV3Strategy.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            asset(),
+            Math.min(aToken.balanceOf(address(this)), _amount),
+            address(this)
+        );
+    }
+
+    function _harvestAndReport()
+    internal
+    override
+    returns (uint256 _totalAssets)
```
Comment (samtvlabs) [inline src/ReserveVault/strategy/AaveV3Strategy.sol:55]: Do we want to make all calls to `lendingPool` nonRentrant ?

**Review by samtvlabs — COMMENTED** [review]
Look good overall ! A couple of clarifications / nits

File: src/ReserveVault/strategy/AaveV3Strategy.sol (hunk 2 above)
Comment (Defcon023) [reply src/ReserveVault/strategy/AaveV3Strategy.sol:55]: Good catch , I have now added a reentrancy guard to all the functions (calling _freeFunds, _deployFunds and _harvestAndReport) in the base strategy 
https://github.com/tvl-labs/khalani-core/pull/200/commits/f3d06864cebeff1feafc4279888ccd3f3de22b3a

File: src/ReserveVault/strategy/AaveV3Strategy.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    ) BaseStrategy(_asset, _name, _symbol) {
+        // Set the aToken based on the asset we are using.
+        lendingPool = _lendingPool;
+        aToken = IERC20(lendingPool.getReserveData(address(_asset)).aTokenAddress);
+
+        // Make sure its a real token.
+        require(address(aToken) != address(0), "!aToken");
+
+        // Make approve the lending pool for cheaper deposits.
+        _asset.safeApprove(address(lendingPool), type(uint256).max);
```
[the remark this answers is not in the document]
Comment (Defcon023) [reply src/ReserveVault/strategy/AaveV3Strategy.sol:32]: Right!, for AaveStrategy this is okay
But for other strategy (we will have to call approve while deploying funds)

**Review by samtvlabs — APPROVED** [review]

**Pull request merged**